File: 020_Valid_Parentheses/020_2.py
class Solution:
    def isValid(self, s):
        """
        :type s: str
        :rtype: bool
        """
        stack = []
        vi = {']':'[', '}':'{', ')':'('}
        for char in s:
            if char in vi.values():
                stack.append(char)
            elif char in vi.keys():
                if stack == [] or stack.pop() != vi[char]:
                    return False
        return stack == []

File: 020_Valid_Parentheses/020_2.cpp
class Solution {
public:
    bool isValid(string s) {
        unordered_map<char, char> map;
        map[')'] = '(';
        map[']'] = '[';
        map['}'] = '{';
        stack<char> stack;
        for (char c : s) {
            if (stack.empty() && map.find(c) != map.end()) {
                return false;
            }
            else if (map.find(c) != map.end() && map[c] == stack.top()) {
                stack.pop();
            }
            else {
                stack.push(c);
            }
        }
        return stack.empty();
    }
};